File: packages/mobile/ios/App/App/NativeAppOpenAdPlugin.swift
import Capacitor
import Foundation
import GoogleMobileAds
import UIKit

@objc(NativeAppOpenAdPlugin)
public class NativeAppOpenAdPlugin: CAPPlugin, CAPBridgedPlugin, FullScreenContentDelegate {
    public let identifier = "NativeAppOpenAdPlugin"
    public let jsName = "NativeAppOpenAd"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "load", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "show", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isLoaded", returnType: CAPPluginReturnPromise)
    ]

    private static let maximumAdAge: TimeInterval = 4 * 60 * 60
    private var appOpenAd: AppOpenAd?
    private var loadedAt: Date?
    private var isLoading = false
    private var isShowing = false
    private var showCall: CAPPluginCall?

    @objc public func load(_ call: CAPPluginCall) {
        Task { @MainActor [weak self] in
            guard let self else {
                call.reject("App Open plugin is unavailable")
                return
            }
            await self.loadOnMainActor(call)
        }
    }

    @MainActor
    private func loadOnMainActor(_ call: CAPPluginCall) async {
        guard let adId = call.getString("adId"), !adId.isEmpty else {
            call.reject("adId is required")
            return
        }

        if hasFreshAd {
            call.resolve()
            return
        }
        clearExpiredAd()

        guard !isLoading else {
            call.reject("An App Open ad is already loading")
            return
        }

        isLoading = true
        do {
            let ad = try await AppOpenAd.load(with: adId, request: Request())
            appOpenAd = ad
            loadedAt = Date()
            isLoading = false
            call.resolve()
        } catch {
            clearAd()
            isLoading = false
            call.reject(error.localizedDescription)
        }
    }

    @objc public func isLoaded(_ call: CAPPluginCall) {
        Task { @MainActor [weak self] in
            guard let self else {
                call.reject("App Open plugin is unavailable")
                return
            }
            self.resolveIsLoaded(call)
        }
    }

    @MainActor
    private func resolveIsLoaded(_ call: CAPPluginCall) {
        clearExpiredAd()
        call.resolve(["value": hasFreshAd])
    }

    @objc public func show(_ call: CAPPluginCall) {
        Task { @MainActor [weak self] in
            guard let self else {
                call.reject("App Open plugin is unavailable")
                return
            }
            self.showOnMainActor(call)
        }
    }

    @MainActor
    private func showOnMainActor(_ call: CAPPluginCall) {
        clearExpiredAd()
        guard let appOpenAd, hasFreshAd, !isShowing else {
            call.reject("App Open ad is not ready")
            return
        }
        guard let viewController = bridge?.viewController else {
            call.reject("No root view controller")
            return
        }

        isShowing = true
        showCall = call
        appOpenAd.fullScreenContentDelegate = self
        appOpenAd.present(from: viewController)
    }

    @MainActor
    public func adWillPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
        notifyListeners("nativeAppOpenAdShown", data: [:])
    }

    @MainActor
    public func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        let call = showCall
        clearAd()
        notifyListeners("nativeAppOpenAdDismissed", data: [:])
        call?.resolve()
    }

    @MainActor
    public func ad(
        _ ad: FullScreenPresentingAd,
        didFailToPresentFullScreenContentWithError error: Error
    ) {
        let call = showCall
        let errorCode = (error as NSError).code
        clearAd()
        notifyListeners("nativeAppOpenAdFailedToShow", data: ["code": errorCode])
        call?.reject(error.localizedDescription)
    }

    @MainActor
    private var hasFreshAd: Bool {
        guard appOpenAd != nil, let loadedAt else { return false }
        return Date().timeIntervalSince(loadedAt) < Self.maximumAdAge
    }

    @MainActor
    private func clearExpiredAd() {
        if !hasFreshAd {
            clearAd()
        }
    }

    @MainActor
    private func clearAd() {
        appOpenAd = nil
        loadedAt = nil
        isShowing = false
        showCall = nil
    }
}

@objc(BridgeViewController)
public class BridgeViewController: CAPBridgeViewController {
    public override func capacitorDidLoad() {
        bridge?.registerPluginInstance(NativeAppOpenAdPlugin())
    }
}

File: packages/mobile/android/app/src/main/java/app/trizum/capacitor/NativeAppOpenAdPlugin.java
package app.trizum.capacitor;

import androidx.annotation.NonNull;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.android.gms.ads.AdError;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.FullScreenContentCallback;
import com.google.android.gms.ads.LoadAdError;
import com.google.android.gms.ads.appopen.AppOpenAd;

@CapacitorPlugin(name = "NativeAppOpenAd")
public class NativeAppOpenAdPlugin extends Plugin {

    private static final long MAX_AD_AGE_MILLISECONDS = 4 * 60 * 60 * 1000L;

    private AppOpenAd appOpenAd;
    private long loadedAt;
    private boolean loading;
    private boolean showing;

    @PluginMethod
    public void load(PluginCall call) {
        String adId = call.getString("adId");
        if (adId == null || adId.isBlank()) {
            call.reject("adId is required");
            return;
        }

        getActivity().runOnUiThread(() -> loadOnMainThread(call, adId));
    }

    @PluginMethod
    public void isLoaded(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            JSObject result = new JSObject();
            result.put("value", hasFreshAd());
            call.resolve(result);
        });
    }

    @PluginMethod
    public void show(PluginCall call) {
        getActivity().runOnUiThread(() -> showOnMainThread(call));
    }

    private void loadOnMainThread(PluginCall call, String adId) {
        if (hasFreshAd()) {
            call.resolve();
            return;
        }
        clearExpiredAd();

        if (loading) {
            call.reject("An App Open ad is already loading");
            return;
        }

        loading = true;
        AppOpenAd.load(
            getContext(),
            adId,
            new AdRequest.Builder().build(),
            new AppOpenAd.AppOpenAdLoadCallback() {
                @Override
                public void onAdLoaded(@NonNull AppOpenAd loadedAd) {
                    appOpenAd = loadedAd;
                    loadedAt = System.currentTimeMillis();
                    loading = false;
                    call.resolve();
                }

                @Override
                public void onAdFailedToLoad(@NonNull LoadAdError error) {
                    appOpenAd = null;
                    loadedAt = 0;
                    loading = false;
                    call.reject(error.getMessage());
                }
            }
        );
    }

    private void showOnMainThread(PluginCall call) {
        if (!hasFreshAd() || showing) {
            clearExpiredAd();
            call.reject("App Open ad is not ready");
            return;
        }

        AppOpenAd ad = appOpenAd;
        if (ad == null) {
            call.reject("App Open ad is not ready");
            return;
        }

        showing = true;
        ad.setFullScreenContentCallback(
            new FullScreenContentCallback() {
                @Override
                public void onAdShowedFullScreenContent() {
                    notifyListeners("nativeAppOpenAdShown", new JSObject());
                }

                @Override
                public void onAdDismissedFullScreenContent() {
                    clearAd();
                    notifyListeners("nativeAppOpenAdDismissed", new JSObject());
                    call.resolve();
                }

                @Override
                public void onAdFailedToShowFullScreenContent(@NonNull AdError error) {
                    clearAd();
                    notifyListeners("nativeAppOpenAdFailedToShow", toError(error));
                    call.reject(error.getMessage());
                }
            }
        );
        ad.show(getActivity());
    }

    private boolean hasFreshAd() {
        return appOpenAd != null && System.currentTimeMillis() - loadedAt < MAX_AD_AGE_MILLISECONDS;
    }

    private void clearExpiredAd() {
        if (!hasFreshAd()) {
            clearAd();
        }
    }

    private void clearAd() {
        appOpenAd = null;
        loadedAt = 0;
        showing = false;
    }

    private JSObject toError(AdError error) {
        JSObject result = new JSObject();
        result.put("code", error.getCode());
        return result;
    }
}
